[DOC-4822] Docs for private clusters (#15342)
cockroachdb/docs: cockroachcloud/private-clusters.md

---
title: Create private CockroachDB Dedicated clusters (Preview)
summary: Learn how to create CockroachDB Dedicated clusters with no node-level public IP addresses.
toc: true
docs_area: manage.security
---

{% include feature-phases/preview-opt-in.md %}

Limiting access to a CockroachDB cluster's nodes over the public internet is an important security practice and is also a compliance requirement for many organizations. {{ site.data.products.dedicated }} private clusters allow organizations to meet this objective.

> By default, {{ site.data.products.db }} has safeguards in place to protect cluster's data from the public internet. Ingress traffic to a cluster is routed through a load balancer, and it's possible to restrict inbound connections using a combination of [IP allowlisting](/docs/cockroachcloud/network-authorization.html#ip-allowlisting) and [AWS PrivateLink](/docs/cockroachcloud/network-authorization.html#aws-privatelink) or [GCP VPC peering](/docs/cockroachcloud/network-authorization.html#vpc-peering). However, data egress operations such as exports, backups, and Change Data Capture (CDC) use public subnets.

Reviewer: Link to exports, backups, and CDC docs in last line.

```suggestion
By default, {{ site.data.products.db }} has safeguards in place to protect your cluster's data from the public internet. Ingress traffic to a cluster is routed through a load balancer, and it is possible to restrict inbound connections using a combination of [IP allowlisting](/docs/cockroachcloud/network-authorization.html#ip-allowlisting) and [AWS PrivateLink](/docs/cockroachcloud/network-authorization.html#aws-privatelink) or [GCP VPC peering](/docs/cockroachcloud/network-authorization.html#vpc-peering). However, data egress operations such as exports, backups, and Change Data Capture (CDC) use public subnets.
```

> On the other hand, a private {{ site.data.products.dedicated }} cluster's nodes have no public IP addresses, and egress traffic moves over private subnets through a highly-available NAT gateway that is unique to the cluster. This page explains what happens when you create a private cluster. Existing clusters that are not created as private clusters can't be switched in place, unless you choose to perform full data migration to a new private cluster.

Reviewer: > Existing clusters that are not created as private clusters can't be switched in place, unless you choose to perform full data migration to a new private cluster.

This line feels tacked on here; perhaps either add a line break or move it below the following callout.

Also (speaking as migration docs writer), what is meant by "full data migration to a new private cluster"? Is that something that we enable, and if so, can we provide a link or hint on how this is accomplished on Dedicated?

Author: I moved this to a new section at the bottom of the page, and added a link to the Migration page.

{{site.data.alerts.callout_info}}
Private clusters are not available with {{ site.data.products.serverless-plan }}.
{{site.data.alerts.end}}

## Create or migrate to a private cluster

To be enrolled in the preview and deploy a new private cluster on AWS or GCP, contact your account team.

After your organization is enrolled in the preview:

> - By default, newly-created {{ site.data.products.dedicated }} clusters deployed on GCP after TODO will be private clusters. Existing clusters deployed on GCP can't be switched in place to be private.

Reviewer: TODO needs to be filled in.

> - By default, newly-created {{ site.data.products.dedicated }} clusters deployed on AWS will **not** be private, and can't be switched in place to be private after the fact. For a cluster on AWS to be private, you can create the cluster using CockroachDB Cloud API by specifying a special field. Contact your account team for details.

Reviewer: See style guide on contractions.

Link to Cloud API docs.

```suggestion
- By default, newly-created {{ site.data.products.dedicated }} clusters deployed on AWS will **not** be private, and cannot be made private after being created. For a cluster on AWS to be private, you can create the cluster using the {{ site.data.products.db }} API by specifying a special field. Contact your account team for details.
```

When you create a private cluster:

> 1. A private subnet is created per requested region.

Reviewer: ```suggestion
1. One private subnet is created per requested region.
```

1. Each node is connected to the regional private subnet.
1. A highly-available NAT gateway is created with static egress public IP addresses. For private clusters deployed on AWS, the NAT gateways are created in three separate availability zones to mitigate against the risk of an availability zone outage.
1. All egress traffic from the cluster nodes to non-cloud buckets is sent across the private subnet and through the NAT gateway to reach its destination.
1. All egress traffic from the cluster nodes to S3 (for private clusters on AWS) or Google Cloud Storage (for private clusters on GCP) is sent across the private subnet and through the private network for the relevant cloud provider (S3 Gateway endpoints on AWS and Private Google Access on GCP).

## Limit incoming connections from egress operations

> Egress traffic from a private cluster will always appear to come from the static IP addresses that comprise the cluster's NAT gateway. To determine the NAT gateway's IP addresses, you can initiate an egress operation, such as an `EXPORT` or `BACKUP` operation on the cluster and observe the source addresses of the resulting connections to your infrastructure. Cockroach Labs recommends that you allow connections to your infrastructure only from those IP addresses.

Reviewer: Link to export and backup docs.

---
title: Create private CockroachDB Dedicated clusters (Preview)
summary: Learn how to create CockroachDB Dedicated clusters with no node-level public IP addresses.
toc: true
docs_area: manage.security
---

{% include feature-phases/preview-opt-in.md %}

Limiting access to a CockroachDB cluster's nodes over the public internet is an important security practice and is also a compliance requirement for many organizations. {{ site.data.products.dedicated }} private clusters allow organizations to meet this objective.

> By default, {{ site.data.products.db }} has safeguards in place to protect cluster's data from the public internet. Ingress traffic to a cluster is routed through a load balancer, and it is possible to restrict inbound connections using a combination of [IP allowlisting](/docs/cockroachcloud/network-authorization.html#ip-allowlisting) and [AWS PrivateLink](/docs/cockroachcloud/network-authorization.html#aws-privatelink) or [GCP VPC peering](/docs/cockroachcloud/network-authorization.html#vpc-peering). However, data egress operations such as [exports](/docs/stable/export.html), [backups](/docs/stable/backup.html), and [Change Data Capture (CDC)](/docs/stable/change-data-capture-overview.html) use public subnets.

Reviewer 2: ```suggestion
By default, {{ site.data.products.db }} has safeguards in place to protect cluster's data from the public internet. Ingress traffic to a cluster is routed through a load balancer, and it is possible to restrict inbound connections using a combination of [IP allowlisting](/docs/cockroachcloud/network-authorization.html#ip-allowlisting), and either of [AWS PrivateLink](/docs/cockroachcloud/network-authorization.html#aws-privatelink) or [GCP VPC peering](/docs/cockroachcloud/network-authorization.html#vpc-peering) depending on your cloud provider. However, data egress operations such as [exports](/docs/stable/export.html), [backups](/docs/stable/backup.html), and [Change Data Capture (CDC)](/docs/stable/change-data-capture-overview.html) use public subnets.
```

> On the other hand, a private {{ site.data.products.dedicated }} cluster's nodes have no public IP addresses, and egress traffic moves over private subnets through a highly-available NAT gateway that is unique to the cluster. This page explains what happens when you create a private cluster.

Reviewer 2: ```suggestion
On the other hand, a private {{ site.data.products.dedicated }} cluster's nodes have no public IP addresses, and egress traffic moves over private subnets and through a highly-available NAT gateway that is unique to the cluster. This page explains what happens when you create a private cluster.
```

## Create a private cluster

To be enrolled in the preview and deploy a new private cluster on AWS or GCP, contact your account team.

After your organization is enrolled in the preview:

- By default, newly-created {{ site.data.products.dedicated }} clusters deployed on GCP will be private clusters.
- By default, newly-created {{ site.data.products.dedicated }} clusters deployed on AWS will **not** be private clusters. To create a private cluster on AWS, you can specify a special field using the [{{ site.data.products.db }} API](/docs/cockroachcloud/cloud-api.html). Contact your account team for details.
- An existing cluster can't be migrated in-place to a private cluster.

When you create a private cluster:

1. One private subnet is created per requested region.
1. Each node is connected to the regional private subnet.
1. A highly-available NAT gateway is created with static egress public IP addresses. For private clusters deployed on AWS, the NAT gateways are created in three separate availability zones to mitigate against the risk of an availability zone outage.

> 1. All egress traffic from the cluster nodes to non-cloud buckets is sent across the private subnet and through the NAT gateway to reach its destination.

Reviewer 2: ```suggestion
1. All egress traffic from the cluster nodes to non-cloud external resources is sent across the private subnet and through the NAT gateway to reach its destination.
```

1. All egress traffic from the cluster nodes to S3 (for private clusters on AWS) or Google Cloud Storage (for private clusters on GCP) is sent across the private subnet and through the private network for the relevant cloud provider (S3 Gateway endpoints on AWS and Private Google Access on GCP).

## Limit inbound connections from egress operations

> Egress traffic from a private cluster will always appear to come from the static IP addresses that comprise the cluster's NAT gateway. To determine the NAT gateway's IP addresses, you can initiate an egress operation such as an [`EXPORT`](/docs/stable/export.html) or [`BACKUP`](/docs/stable/backup.html) operation on the cluster and observe the source addresses of the resulting connections to your infrastructure. Cockroach Labs recommends that you allow connections to your infrastructure only from those IP addresses.

Reviewer 2: ```suggestion
Egress traffic from a private cluster to non-cloud external resources will always appear to come from the static IP addresses that comprise the cluster's NAT gateway. To determine the NAT gateway's IP addresses, you can initiate an egress operation such as an [`EXPORT`](/docs/stable/export.html) or [`BACKUP`](/docs/stable/backup.html) operation on the cluster and observe the source addresses of the resulting connections to your non-cloud external resources. Cockroach Labs recommends that you allow connections to such resources only from those IP addresses.
```

{{ site.data.products.db }} has safeguards in place to protect cluster's data from the public internet. Ingress traffic to a cluster is routed through a load balancer, and it's possible to restrict inbound connections using IP allow-lists, or Virtual Private Cloud (VPC) peering. However, data egress operations such as exports, backups, and Change Data Capture (CDC) use public subnets.

A private {{ site.data.products.dedicated }} cluster's nodes have no public IP addresses, and egress traffic moves over private subnets through a highly-available NAT gateway that is unique to the cluster. This page explains what happens when you create a private cluster, migrate an existing cluster to a private cluster, or migrate a private cluster to no longer be private.

{{site.data.alerts.callout_info}}
Private clusters are not available with {{ site.data.products.serverless-plan }}.
{{site.data.alerts.end}}

## Create or migrate to a private cluster

To be enrolled in the preview, to deploy a new private cluster on AWS, or to migrate an existing cluster to a private cluster, contact your account team.

After your organization is enrolled in the preview:

> - By default, newly-created {{ site.data.products.dedicated }} clusters deployed on GCP after TODO will be private clusters, but can be migrated to private clusters.

Reviewer 2: ```suggestion
- By default, newly-created {{ site.data.products.dedicated }} clusters deployed on GCP after TODO will be private clusters. Existing clusters deployed on GCP can't be switched in place to be private. 
```

> - By default, newly-created {{ site.data.products.dedicated }} clusters deployed on AWS will **not** be private clusters, but can be migrated to private clusters.

Reviewer 2: ```suggestion
- By default, newly-created {{ site.data.products.dedicated }} clusters deployed on AWS will **not** be private, and can't be switched in place to be private after the fact. For a cluster on AWS to be private, please create the cluster using CockroachDB Cloud API while specifying a special field (details can be shared by your account team).
```

When you create a private cluster:

> 1. A private subnet is created.

Reviewer 2: ```suggestion
1. A private subnet is created per requested region.
```

> 1. As each cluster node is created, it is connect to the private subnet.

Reviewer 2: ```suggestion
1. Each cluster node is created in the regional private subnet.
```

> 1. A NAT gateway is created which spans three separate availability zones, to mitigate against the risk of an availability zone outage.

Reviewer 2: ```suggestion
1. A highly-available NAT gateway is created with static egress public IP. In case of AWS, the NAT gateways are created in three separate availability zones to mitigate against the risk of an availability zone outage.
```

> 1. All egress traffic from the cluster is sent across the private subnet and the NAT gateway to reach its destination.

Reviewer 2: ```suggestion
1. All egress traffic from the cluster nodes to non-cloud buckets is sent across the private subnet and through the NAT gateway to reach its destination.
```

> 

Reviewer 2: Add another point here, something like:
1. All egress traffic from the cluster nodes to cloud buckets (S3 for clusters on AWS, and GCS for clusters on GCP) is sent across the private subnet and through the cloud provider-specific private network. Clusters deployed on AWS use [S3 Gateway endpoints](https://docs.aws.amazon.com/vpc/latest/privatelink/vpc-endpoints-s3.html), while ones on GCP use [Private Google Access](https://cloud.google.com/vpc/docs/private-google-access) to achieve the egress private connectivity.

> When you migrate an existing cluster to be private, the following sequence occurs with no downtime:

Reviewer 2: We don't support migration or to be specific in-place switch to private yet. We should remove the migration specific sections below. For an existing cluster to be private, the only way is to create a new cluster and migrate data to it.

1. A private subnet is created, in addition to the public subnet each cluster node is connected to.
1. One node at a time is drained and restarts, but is now connected only to the private subnet.
1. A NAT gateway is created which spans three separate availability zones, to mitigate against the risk of an availability zone outage.
1. All egress traffic from the cluster is sent across the private subnet and the NAT gateway to reach its destination.

When you migrate an existing private cluster to no longer be private, the following sequence occurs with no downtime:

1. A public subnet is created, in addition to the private subnet each cluster node is connected to.
1. One node at a time is drained and restarts, but is now connected only to the public subnet.
1. The NAT gateway is destroyed.
1. All egress traffic from the cluster is sent across the public subnet to reach its destination.

## Limit incoming connections from egress operations

> Egress traffic from a private cluster will always appear to come from the three IP addresses that comprise the cluster's NAT gateway. To determine the NAT gateway's IP addresses, you can initiate an egress operation, such as an `EXPORT` or `BACKUP` operation, on the cluster and observe the source addresses of the resulting connections to your infrastructure. Cockroach Labs recommends that you allow connections to your infrastructure only from those IP addresses.

Reviewer 2: ```suggestion
Egress traffic from a private cluster will always appear to come from the static IP addresses that comprise the cluster's NAT gateway. To determine the NAT gateway's IP addresses, you can initiate an egress operation, such as an `EXPORT` or `BACKUP` operation on the cluster and observe the source addresses of the resulting connections to your infrastructure. Cockroach Labs recommends that you allow connections to your infrastructure only from those IP addresses.
```

> If you migrate an existing cluster to or from a private cluster, you may need to make adjustments to your network connection rules.

Reviewer 2: Remove as migration / switch in place is not supported.
